packages/store/src/external/hydrate.js

```diff
@@ -0,0 +1,115 @@
+// @ts-check
+
+import '../types';
+
+import { makeWeakStore } from '../weak-store';
+import { makeStore } from '../store';
+
+/**
+ * @callback MakeBackingStore
+ * @param {HydrateHook} hydrateHook
+ * @returns {BackingStore}
+ */
+
+/**
+ * This creates an external store maker for a given storage backend, supporting
+ * the Closure interface that the rewriter targets.
+ *
+ * @template {Array<any>} A
+ * @template {ExternalInstance} T
+ * @param {MakeBackingStore} makeBackingStore
+ * @returns {MakeHydrateExternalStore<A, T>}
+ */
+export const makeHydrateExternalStoreMaker = makeBackingStore => {
+  const serialize = JSON.stringify;
+  const unserialize = JSON.parse;
+
+  /** @type {WeakStore<T, [string, string]>} */
+  const instanceToKey = makeWeakStore('instance');
+
+  let lastStoreKey = 0;
+
+  // This has to be a strong store, since it is indexed by key.
+  const storeKeyToHydrate = makeStore('storeKey');
+
+  /**
+   * Create a data object that queues writes to the store.
+   *
+   * @param {HydrateData} data
+   * @param {() => void} markDirty
+   */
+  const makeActiveData = (data, markDirty) => {
+    const activeData = {};
+    // For every property in data...
+    for (const prop of Object.getOwnPropertyNames(data)) {
+      // Define a getter and setter on activeData.
+      Object.defineProperty(activeData, prop, {
+        get: () => data[prop],
+        set: value => {
+          data[prop] = value;
+          markDirty();
+        },
+      });
+    }
+    return harden(activeData);
+  };
+
+  /**
+   * @type {BackingStore}
+   */
+  let backing;
+  const hydrateHook = {
+    getKey(value) {
+      return instanceToKey.get(value);
+    },
+    load([storeKey, instanceKey]) {
+      const hydrate = storeKeyToHydrate.get(storeKey);
+      const store = backing.findStore(storeKey);
+
+      const data = unserialize(store.get(instanceKey));
+      const markDirty = () => store.set(instanceKey, serialize(data));
```

Comment:
I don't think this is a practical implementation of `markDirty`. OTOH, this code itself is not in a position to determine what happens at end-of-transaction. But `makeBackingStore` is. Here, can we tell `store` simply that the `instanceKey` is dirty, combined with adding the `serialize` function as an argument to `backing.makeStore`, so the backing store can be told here what will need to be serialized, and how to serialize it, but it leave it to the backing store to determine when to do so.

Which btw raises the SwingSet question of how the backing store will arrange to commit or abort atomically with the vat using it. I also have questions about what happens to backing store state during replay. I think the answer is nothing, since the backing store is a device. All answers from the backing store would be part of the vat's replay log, and the "real" backing store would not be reconnected until the vat replayed to that point. But that point does not include the writes from the aborted crank, so that will need to be absent from the committed backing store state. Thus, `makeBackingStore` must already be aware of transaction boundaries, and so can know when to serialize dirty state.

Reply:
> I don't think this is a practical implementation of `markDirty`. We should serialize dirty data only once per transaction.

Okay, I will add the necessary hooks to do that.

I think the serialization should only happen in the backing store, not this layer.  So I will push it down.